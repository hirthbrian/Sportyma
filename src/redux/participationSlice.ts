import { createSlice } from "@reduxjs/toolkit";
import type { PayloadAction } from "@reduxjs/toolkit";
import type { RootState } from "./store";

import trainig1 from "../../data/training_1.json";
import trainig2 from "../../data/training_2.json";

// Define a type for the slice state
interface ParticipationState {
  [key: number]: string | null;
}

// Define the initial state using that type
const initialState: ParticipationState = {
  1: trainig1.participation,
  2: trainig2.participation,
};

export const participationSlice = createSlice({
  name: "participation",
  // `createSlice` will infer the state type from the `initialState` argument
  initialState,
  reducers: {
    approve: (state, action: PayloadAction<number>) => {
      state[action.payload] = "PRESENT";
    },
    refuse: (state, action: PayloadAction<number>) => {
      state[action.payload] = "ABSENT";
    },
  },
});

export const { approve, refuse } = participationSlice.actions;

// Other code such as selectors can use the imported `RootState` type
export const selectParticipationStatus = (state: RootState) =>
  state.participation;

export default participationSlice.reducer;
