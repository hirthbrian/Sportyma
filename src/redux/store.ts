import { configureStore } from "@reduxjs/toolkit";
import participation from "./participationSlice";

export const store = configureStore({
  reducer: {
    participation,
  },
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
