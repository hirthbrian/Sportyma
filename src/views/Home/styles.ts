import styled from "styled-components/native";

import { Colors } from "../../types";

export const Container = styled.FlatList`
  flex: 1;
  padding: 20px;
  background-color: ${Colors.White};
`;
