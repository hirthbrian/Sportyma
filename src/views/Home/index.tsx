import { useNavigation } from "@react-navigation/native";
import { useAppSelector } from "../../redux/hook";

import { ParticipationButton } from "../../molecules";
import { Separator } from "../../atoms";
import { loadEventData } from "../../utils";

import { Container } from "./styles";

const Home = () => {
  const participationStatus = useAppSelector((state) => state.participation);
  const navigation = useNavigation();
  const data = loadEventData();

  const getActiveColor = (status: string) => {
    if (status === "PRESENT") {
      return "#29D48C";
    } else if (status === "ABSENT") {
      return "#EF4948";
    }
  };

  const renderItem = ({ item: { title, id } }) => {
    const status = participationStatus[id];
    return (
      <ParticipationButton
        title={title}
        activeColor={getActiveColor(status)}
        isActive={status === "PRESENT" || status === "ABSENT"}
        onPress={() => navigation.navigate("Event", { id })}
      />
    );
  };

  return (
    <Container
      data={data}
      renderItem={renderItem}
      ItemSeparatorComponent={() => <Separator />}
    />
  );
};

export default Home;
