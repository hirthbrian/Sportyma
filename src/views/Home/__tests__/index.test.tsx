import React from "react";
import renderer from "react-test-renderer";
import { Provider } from "react-redux";

import { store } from "../../../redux/store";

import Home from "..";

const mockedNavigate = jest.fn();

jest.mock("@react-navigation/native", () => {
  const actualNav = jest.requireActual("@react-navigation/native");
  return {
    ...actualNav,
    useNavigation: () => ({
      navigate: mockedNavigate,
    }),
  };
});

describe("<Home />", () => {
  let props;

  beforeEach(() => {
    props = {};
  });

  it("renders correcly", () => {
    expect(
      renderer
        .create(
          <Provider store={store}>
            <Home {...props} />
          </Provider>
        )
        .toJSON()
    ).toBeTruthy();
  });
  it("match snapshot", () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <Home {...props} />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
