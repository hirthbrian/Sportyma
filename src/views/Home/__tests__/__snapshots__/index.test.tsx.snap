// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home /> match snapshot 1`] = `
<RCTScrollView
  ItemSeparatorComponent={[Function]}
  data={
    Array [
      Object {
        "enddate": "2023-02-16T20:35:00+00:00",
        "id": 1,
        "information": "N’oubliez pas vos lorem ipsum dolor sit amet, consectetur adipiscing elit.",
        "location": Object {
          "address": "12 rue de l’Extenwoerth",
          "city": "67100 Strasbourg",
          "place": "Stade de la Meinau",
        },
        "participation": null,
        "startdate": "2023-02-16T20:35:00+00:00",
        "title": "Entraînement à Strasbourg",
      },
      Object {
        "enddate": "2023-02-16T20:35:00+00:00",
        "id": 2,
        "information": null,
        "location": Object {
          "address": "24 Rue du Commandant Guilbaud",
          "city": "75016 Paris",
          "place": "Le Parc des Princes",
        },
        "participation": "ABSENT",
        "startdate": "2023-02-08T19:35:00+00:00",
        "title": "Entraînement à Paris",
      },
    ]
  }
  getItem={[Function]}
  getItemCount={[Function]}
  keyExtractor={[Function]}
  onContentSizeChange={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  style={
    Array [
      Object {
        "backgroundColor": "#FFFFFF",
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
        "paddingBottom": 20,
        "paddingLeft": 20,
        "paddingRight": 20,
        "paddingTop": 20,
      },
    ]
  }
  viewabilityConfigCallbackPairs={Array []}
>
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "flexBasis": 0,
              "flexGrow": 1,
              "flexShrink": 1,
            },
          ]
        }
      >
        <View
          style={
            Object {
              "transform": Array [
                Object {
                  "scale": 1,
                },
              ],
            }
          }
        >
          <View
            isActive={false}
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#FFFFFF",
                  "borderColor": "#2559f233",
                  "borderRadius": 50,
                  "borderWidth": 2,
                  "flexDirection": "row",
                  "height": 50,
                  "justifyContent": "center",
                },
              ]
            }
          >
            <Image
              isActive={false}
              style={
                Array [
                  Object {
                    "height": 18,
                    "marginRight": 10,
                    "tintColor": "#2559F2",
                    "width": 18,
                  },
                ]
              }
            />
            <Text
              isActive={false}
              size={13}
              style={
                Array [
                  Object {
                    "color": "#2559F2",
                    "fontFamily": "Kanit-Medium",
                    "fontSize": 13,
                    "letterSpacing": 1,
                  },
                ]
              }
              weight={500}
            >
              ENTRAÎNEMENT À STRASBOURG
            </Text>
          </View>
        </View>
      </View>
      <View
        style={
          Array [
            Object {
              "backgroundColor": "#DFE5EC",
              "height": 1,
              "lineHeight": 28,
              "marginBottom": 15,
              "marginLeft": 0,
              "marginRight": 0,
              "marginTop": 15,
            },
          ]
        }
      />
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "flexBasis": 0,
              "flexGrow": 1,
              "flexShrink": 1,
            },
          ]
        }
      >
        <View
          style={
            Object {
              "transform": Array [
                Object {
                  "scale": 1,
                },
              ],
            }
          }
        >
          <View
            color="#EF4948"
            isActive={true}
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#EF4948",
                  "borderColor": "#EF4948",
                  "borderRadius": 50,
                  "borderWidth": 2,
                  "flexDirection": "row",
                  "height": 50,
                  "justifyContent": "center",
                },
              ]
            }
          >
            <Image
              color="#EF4948"
              isActive={true}
              style={
                Array [
                  Object {
                    "height": 18,
                    "marginRight": 10,
                    "tintColor": "#FFFFFF",
                    "width": 18,
                  },
                ]
              }
            />
            <Text
              isActive={true}
              size={13}
              style={
                Array [
                  Object {
                    "color": "#FFFFFF",
                    "fontFamily": "Kanit-Medium",
                    "fontSize": 13,
                    "letterSpacing": 1,
                  },
                ]
              }
              weight={500}
            >
              ENTRAÎNEMENT À PARIS
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
